File: src/test/java/ez/spring/vertx/web/SpringVertxWebTests.java
package ez.spring.vertx.web;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.annotation.Timed;
import org.springframework.test.context.junit4.SpringRunner;

import ez.spring.vertx.EzJob;
import io.vertx.core.Future;
import io.vertx.core.Vertx;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.core.json.Json;
import io.vertx.ext.web.client.HttpResponse;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.WebClientOptions;

@RunWith(SpringRunner.class)
@SpringBootTest
public class SpringVertxWebTests {
  private Logger log = LoggerFactory.getLogger(getClass());
  @Autowired
  private HttpServerOptions httpServerOptions;
  private Vertx vertx;
  private WebClient webClient;

  @Before
  public void start() {
    vertx = Vertx.vertx();
    webClient = WebClient.create(vertx, new WebClientOptions()
      .setDefaultHost("localhost")
      .setDefaultPort(httpServerOptions.getPort())
    );
  }

  @Timed(millis = 2000)
  @Test
  public void checkTestWebHandler() {
    TestWebHandler.Data req = new TestWebHandler.Data();
    req.setValue(String.valueOf(Math.random()));
    TestWebHandler.Data resp = EzJob.create(vertx, "send request to TestWebHandler").<HttpResponse<Buffer>>then(
      p -> webClient.post("/").sendJson(req, p)
    ).thenCompose(r -> Future.succeededFuture(
      Json.decodeValue(r.body(), TestWebHandler.Data.class)
    )).join();
    Assert.assertEquals(req.getValue(), resp.getValue());
  }

  @After
  public void end() {
    webClient.close();
    vertx.close();
  }
}

File: src/test/java/ez/spring/vertx/web/SpringVertxWebTests.kt
package ez.spring.vertx.web

import ez.spring.vertx.http.HttpServerConfiguration
import ez.spring.vertx.web.TestWebHandler.Body
import io.vertx.core.Vertx
import io.vertx.core.http.HttpHeaders
import io.vertx.ext.web.client.WebClient
import io.vertx.ext.web.client.WebClientOptions
import io.vertx.kotlin.ext.web.client.sendJsonAwait
import kotlinx.coroutines.*
import org.junit.After
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.slf4j.LoggerFactory
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit4.SpringRunner

@RunWith(SpringRunner::class)
@SpringBootTest
class SpringVertxWebTests {
    private val log = LoggerFactory.getLogger(javaClass)
    private val vertx: Vertx = Vertx.vertx()
    private val webClient: WebClient = WebClient.create(
            vertx, WebClientOptions().setDefaultHost("localhost").setDefaultPort(HttpServerConfiguration.DEFAULT_PORT)
    )

    @Test
    fun checkTestWebHandler() {
        val list = mutableListOf<Job>()
        repeat(3) {
            list += GlobalScope.launch {
                val body = Body()
                val v = Math.random() * 1000
                val a = v.toInt()
                val b = v.toString()
                body.a = a
                body.b = b
                val res = webClient.post("/").sendJsonAwait(body)
                Assert.assertEquals(
                        "application/json;charset=utf-8",
                        res.getHeader(HttpHeaders.CONTENT_TYPE.toString())
                )
                val resBody: Body = res.bodyAsJson(Body::class.java)
                Assert.assertEquals(resBody.a.toLong(), a.toLong())
                Assert.assertEquals(resBody.b, b)
            }
        }
        runBlocking {
            list.joinAll()
        }
    }

    @After
    fun end() {
        webClient.close()
    }
}